File: app/src/main/java/org/ole/planet/myplanet/ui/feedback/FeedbackListFragment.java
package org.ole.planet.myplanet.ui.feedback;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import org.ole.planet.myplanet.databinding.FragmentFeedbackListBinding;
import org.ole.planet.myplanet.datamanager.DatabaseService;
import org.ole.planet.myplanet.model.RealmFeedback;
import org.ole.planet.myplanet.model.RealmUserModel;
import org.ole.planet.myplanet.service.UserProfileDbHandler;

import java.util.List;

import io.realm.Realm;
import io.realm.RealmResults;

public class FeedbackListFragment extends Fragment implements FeedbackFragment.OnFeedbackSubmittedListener {
    private FragmentFeedbackListBinding fragmentFeedbackListBinding;
    Realm mRealm;
    RealmUserModel userModel;

    public FeedbackListFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        fragmentFeedbackListBinding = FragmentFeedbackListBinding.inflate(inflater, container, false);
        mRealm = new DatabaseService(getActivity()).getRealmInstance();
        userModel = new UserProfileDbHandler(getActivity()).getUserModel();
        fragmentFeedbackListBinding.fab.setOnClickListener(vi -> {
            FeedbackFragment feedbackFragment = new FeedbackFragment();
            feedbackFragment.setOnFeedbackSubmittedListener(this);
            feedbackFragment.show(getChildFragmentManager(), "");
        });
        return fragmentFeedbackListBinding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        fragmentFeedbackListBinding.rvFeedback.setLayoutManager(new LinearLayoutManager(getActivity()));
        List<RealmFeedback> list = mRealm.where(RealmFeedback.class).equalTo("owner", userModel.name).findAll();
        if (userModel.isManager()) list = mRealm.where(RealmFeedback.class).findAll();
        AdapterFeedback adapterFeedback = new AdapterFeedback(getActivity(), list);
        fragmentFeedbackListBinding.rvFeedback.setAdapter(adapterFeedback);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (!mRealm.isClosed()) mRealm.close();
    }

    @Override
    public void onFeedbackSubmitted() {
        mRealm.executeTransactionAsync(realm -> {}, () -> {
            RealmResults<RealmFeedback> updatedList = mRealm.where(RealmFeedback.class).equalTo("owner", userModel.name).findAll();
            if (userModel.isManager()) updatedList = mRealm.where(RealmFeedback.class).findAll();
            AdapterFeedback adapterFeedback = new AdapterFeedback(getActivity(), updatedList);
            fragmentFeedbackListBinding.rvFeedback.setAdapter(adapterFeedback);
            adapterFeedback.notifyDataSetChanged();
        });
    }
}

File: app/src/main/java/org/ole/planet/myplanet/ui/feedback/FeedbackListFragment.kt
package org.ole.planet.myplanet.ui.feedback

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import dagger.hilt.android.AndroidEntryPoint
import org.ole.planet.myplanet.base.BaseBindingFragment
import org.ole.planet.myplanet.base.BaseRecyclerFragment.Companion.showNoData
import org.ole.planet.myplanet.callback.OnChangedListener
import org.ole.planet.myplanet.databinding.FragmentFeedbackListBinding
import org.ole.planet.myplanet.model.Feedback
import org.ole.planet.myplanet.utils.collectWhenStarted

@AndroidEntryPoint
class FeedbackListFragment : BaseBindingFragment<FragmentFeedbackListBinding>(FragmentFeedbackListBinding::inflate), OnChangedListener {

    private val viewModel: FeedbackListViewModel by viewModels()

    private lateinit var feedbackAdapter: FeedbackAdapter

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        binding.fab.setOnClickListener {
            val feedbackFragment = FeedbackFragment()
            feedbackFragment.setOnFeedbackSubmittedListener(this)
            if (!childFragmentManager.isStateSaved) {
                feedbackFragment.show(childFragmentManager, "")
            }
        }
        return view
    }

    private fun refreshFeedbackListData() {
        viewModel.refreshFeedback()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        feedbackAdapter = FeedbackAdapter()
        binding.rvFeedback.layoutManager = LinearLayoutManager(activity)
        binding.rvFeedback.adapter = feedbackAdapter
        observeFeedbackList()
    }

    private fun observeFeedbackList() {
        collectWhenStarted(viewModel.feedbackList) { feedbackList ->
            updatedFeedbackList(feedbackList)
        }
    }

    override fun onChanged() {
        refreshFeedbackListData()
    }

    private fun updatedFeedbackList(updatedList: List<Feedback>?) {
        if (_binding == null) return
        feedbackAdapter.submitList(updatedList) {
            binding.rvFeedback.scrollToPosition(0)
        }
        val itemCount = updatedList?.size ?: 0
        showNoData(binding.tvMessage, itemCount, "feedback")
        updateTextViewsVisibility(itemCount)
    }

    private fun updateTextViewsVisibility(itemCount: Int) {
        val visibility = if (itemCount == 0) View.GONE else View.VISIBLE
        binding.tvTitle.visibility = visibility
        binding.tvType.visibility = visibility
        binding.tvPriority.visibility = visibility
        binding.tvStatus.visibility = visibility
        binding.tvOpenDate.visibility = visibility
    }
}
